Addressing doc bugs and enhancements in Environments.md (#5260)
microsoft/vscode-docs: docs/python/environments.md

TOCTitle: Environments
ContentId: 8fe4ca8b-fc70-4216-86c7-2c11b6c14cc6
PageTitle: Using Python Environments in Visual Studio Code
DateApproved: 03/20/2022
MetaDescription: Configuring Python Environments in Visual Studio Code
MetaSocialImage: images/tutorial/social.png
---
# Using Python environments in VS Code

This article discusses the helpful Python environments features available in VS Code. An "environment" in Python is the context in which a Python program runs and consists of an interpreter and any number of installed packages.

## Python environments

### Global environments

> By default, any Python interpreter installed runs in its own **global environment**. They aren't specific to a particular project. For example, if you just run `python` (Windows) or `python3` (macOS/Linux) at a new command prompt, you're running in that interpreter's global environment. If Python is not added to PATH, run `py` (Windows) at a new command prompt instead. Any packages that you install or uninstall affect the global environment and all programs that you run within it.

Reviewer: It can also be just `python3` on Windows as well.

```suggestion
By default, any Python interpreter installed runs in its own **global environment**. They aren't specific to a particular project. For example, if you just run `python`, `python3`, or `py` at a new terminal (depending on how you installed Python), you're running in that interpreter's global environment. Any packages that you install or uninstall affect the global environment and all programs that you run within it.
```

> Working in the global environment is an easy way to get started. If you install packages in that environment, in time it will become crowded, making it difficult to thoroughly test an application.

Reviewer: ```suggestion
Working in the global environment is an easy way to get started. If you install packages in that environment, though, in time it will become crowded and make it difficult to properly test an application.
```

### Virtual environments

> To prevent such clutter, developers often create a **virtual environment** for a project. A virtual environment is a subfolder in a project that contains a copy of a specific interpreter. When you activate the virtual environment, any packages you install are installed only in that subfolder. When you then run a Python program within that environment, you know that it's running against only those specific packages. You might need to specify the Python interpreter to use the terminal for installing packages to the global environment. Consider this option if you're not using a **virtual environment**, have multiple versions of Python installed, and are set in the `path` environment variable.

Reviewer: ```suggestion
To prevent such clutter, developers often create a **virtual environment** for a project. A virtual environment is a folder that contains a copy (or symlink) of a specific interpreter. When you install into a virtual environment, any packages you install are installed only in that subfolder. When you then run a Python program within that environment, you know that it's running against only those specific packages.
```

> **Note**: While it's possible to open a virtual environment folder as a workspace, doing so is not recommended and might cause issues with using the Python extension.

### Python environment tools
Once you activate your virtual environment, you’ll need to identify how to manage it and its accompanying packages. The following table explains how to use these Python environments:

|   Tool   | Definition and Purpose |
|   :---:  | --- |

> |   pip    | The Python package manager that installs and updates packages. It's installed with Python 3.4+ by default.      |

Reviewer: ```suggestion
|   pip    | The Python package manager that installs and updates packages. It's installed with Python 3.4+ by default (intall `python3-pip` on Debian-based OSs).      |
```

> |   venv   | Allows you to manage separate package installations for different projects and is installed with Python 3 by default. For Python 2, **virtualenv** would need to be installed manually. |

Reviewer: ```suggestion
|   venv   | Allows you to manage separate package installations for different projects and is installed with Python 3 by default (install `python3-venv` if you are using a Debian-based OS) |
```

|   conda  | Installed with **Anaconda** and **Miniconda**. It can be used to manage both packages and virtual environments. Generally used for data science projects.   |

### Conda environments

> A **conda environment** is a Python environment that's managed using the `conda` package manager (see [Getting started with conda](https://conda.io/projects/conda/en/latest/user-guide/getting-started.html) (conda.io)). Conda works well to create environments with interrelated dependencies as well as binary packages. Unlike virtual environments, which are scoped to a project, conda environments are available globally on any given computer. This availability makes it easy to configure several distinct conda environments and then choose the appropriate one for any given project.

Reviewer: ```suggestion
A **conda environment** is a Python environment that's managed using the `conda` package manager (see [Getting started with conda](https://conda.io/projects/conda/en/latest/user-guide/getting-started.html) (conda.io)). Whether to use a conda environment or a virtual one will depend on your packaging needs, what your team has standardized on, etc.
```

TOCTitle: Environments
ContentId: 8fe4ca8b-fc70-4216-86c7-2c11b6c14cc6
PageTitle: Using Python Environments in Visual Studio Code
DateApproved: 03/20/2022
MetaDescription: Configuring Python Environments in Visual Studio Code
MetaSocialImage: images/tutorial/social.png
---
# Using Python environments in VS Code

This article discusses the helpful Python environments features available in VS Code. An "environment" in Python is the context in which a Python program runs and consists of an interpreter and any number of installed packages.

> If you're new to working with Python environments, you can learn more at [Getting started](#getting-started).

Reviewer 2: I'm concerned if some people will feel like they're being directed out of this doc if they're trying to learn about Python environments, as the getting started section only has links to other docs and tutorials?

Author: That's a good point. Maybe the python-tutorial.md located in the Getting Started section at the bottom will be enough. If I notice any feedback from the community, regarding Python environments, I can add links to the bottom later!

## Python environments

### Global environments

> By default, any Python interpreter installed runs in its own **global environment**. They aren't specific to any one project. For example, if you just run `python` (Windows) or `python3` (macOS/Linux) at a new command prompt, you're running in that interpreter's global environment. Any packages that you install or uninstall affect the global environment and all programs that you run within it.

Reviewer 2: "They aren't specific to a particular project", maybe?
it may also be worth including "py" as an option for Windows for those who forget to add "python" to PATH when installing Python in their machines 😅

Author: That's a really great suggestion! I've added the `py` command and changed the first sentence.

> Working in the global environment is an easy way to get started. In time, that environment will become crowded with the different packages, making it difficult to thoroughly test an application.

Reviewer 2: maybe "But if you install packages in that environment, in time it will become crowded, making it difficult..."

### Virtual environments

> To prevent such clutter, developers often create a **virtual environment** for a project. A virtual environment is a subfolder in a project that contains a copy of a specific interpreter. When you activate the virtual environment, any packages you install are installed only in that environment's subfolder. When you then run a Python program within that environment, you know that it's running against only those specific packages. Be aware that if you're not using a **virtual environment**, and you have multiple versions of Python installed and set in the `path` environment variable, you might need to specify the Python interpreter to use in the terminal for installing packages to the global environment.

Reviewer 2: "in that environment's subfolder" could make it sound as if they subfolder is created inside the global environment. Maybe remove the "environment's" from that phrase? 

I think the last phrase could be improved to be clearer as well, but I don't know how to make it better 😅

Author: I broke up the last phrase a bit. Let me know what you think of it!

> **Note**: While it's possible to open a virtual environment folder as a workspace, doing so is not recommended and might cause issues with using the Python extension.

### Python environment tools
Once you activate your virtual environment, you’ll need to identify how to manage it and its accompanying packages. The following table explains how to use these Python environments:

|   Tool         | Definition and Purpose    |
|   ---          | --- |

> |   pip          | Manages packages and is installed with Python by default.      |

Reviewer 2: not all Python environments come with pip AFAIK

Author: Yes, thanks so  for the callout! I updated this -- I believe pip is installed with newer versions of Python (3.4+).

> |   virtualenv   | Manages virtual environments and is installed with Python by default.   |

Reviewer 2: I don't think virtualenv is installed with Python by default. venv is installed in a lot of them but also not all, AFAIK

Author: Updated this as well. Let me know if you think this table has the potential to provide users some clarity. I'm also happy to remove if not!

- Virtual environments located in a `~/.virtualenvs` folder for [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/).
- Interpreters installed by [pyenv](https://github.com/pyenv/pyenv), [Pipenv](https://pypi.org/project/pipenv/), and [Poetry](https://poetry.eustace.io/).
- Virtual environments located in the path identified by `WORKON_HOME` (as used by [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/)).
- Conda environments that contain a Python interpreter. VS Code doesn't show conda environments that don't contain an interpreter.
- Interpreters installed in a `.direnv` folder for [direnv](https://direnv.net/) under the workspace (project) folder.

> You can also [manually specify an interpreter](#manually-specify-an-interpreter) if Visual Studio Code doesn't locate it automatically.

Reviewer: ```suggestion
You can also [manually specify an interpreter](#manually-specify-an-interpreter) if Visual Studio Code doesn't locate your interpreter automatically.
```

- Virtual environments located in a `~/.virtualenvs` folder for [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/).
- Interpreters installed by [pyenv](https://github.com/pyenv/pyenv), [Pipenv](https://pypi.org/project/pipenv/), and [Poetry](https://poetry.eustace.io/).
- Virtual environments located in the path identified by `WORKON_HOME` (as used by [virtualenvwrapper](https://virtualenvwrapper.readthedocs.io/)).

> - Conda environments that contain a Python interpreter. VS Code doesn't show conda environments that don't contain an interpreter.

Reviewer 2: I believe this will change in the next release, right @karrtikr? We now install a Python interpreter on conda environments that don't come with one?

Reviewer: Correct.
```suggestion
- Conda environments found by `conda env list`. Conda environments which do not have an interpreter will have one installed for them upon selection.
```

### Create a conda environment

> The Python extension automatically detects existing conda environments if the environment contains a Python interpreter. For example, the following command creates a conda environment with the Python 3.4 interpreter and several libraries, which VS Code then shows in the list of available interpreters:

Reviewer: ```suggestion
The Python extension automatically detects existing conda environments. We recommend you install a Python interpreter into your conda environment, otherwise one will be installed for you after you select the environment. For example, the following command creates a conda environment with the Python 3.4 interpreter and several libraries, which VS Code then shows in the list of available interpreters:
```

```bash

> conda create -n env-01 python=3.4 scipy=0.15.0 astroid babel

Reviewer 3: We should update the Python here (say to `3.9`) as we don't officially support any python which has reached EOL: https://endoflife.date/python.

> conda create -n env-01 python=3.4 scipy=0.15.0 astroid babel

Reviewer 3: Also in the additional note section:

> If you create a new conda environment while VS Code is running, use the **Reload Window** command to refresh the environment list shown with **Python: Select Interpreter**; otherwise you may not see the environment there. It might take a short time to appear; if you don't see it at first, wait 15 seconds then try using the command again

Users can now use the refresh icon on the top right of the interpreters list instead of reloading:

![image](https://user-images.githubusercontent.com/13199757/162548522-d02d3950-c785-42c8-8043-37c3b8170704.png)

So we can possibly point to that instead. cc/ @luabud Although that won't detect any new global installs.
